geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/tms/TileReader.scala

```diff
@@ -24,18 +24,46 @@ trait TileReader {
 
 object TileReaders {
 
+  private def rezoom(zoom: Int, x: Int, y: Int, maxZoom: Int, read: SpatialKey => Tile) = {
+    val dz = zoom - maxZoom
+    val key = SpatialKey((x / math.pow(2, dz)).toInt, (y / math.pow(2, dz)).toInt)
+    val dx = x - key._1 * math.pow(2, dz).toInt
+    val dy = y - key._2 * math.pow(2, dz).toInt
+    val tile = read(key)
+    val w = tile.cols
+    val h = tile.rows
+    val tw = (w / math.pow(2, dz)).toInt
+    val th = (h / math.pow(2, dz)).toInt
+    val x0 = tw * dx
+    val x1 = tw * (dx + 1)
+    val y0 = th * dy
+    val y1 = th * (dy + 1)
+    tile.crop(x0, y0, x1, y1).resample(w, h)
+  }
+
   private class CatalogTileReader(
     valueReader: ValueReader[LayerId],
-    catalog: String
+    catalog: String,
+    overzooming: Boolean
   ) extends TileReader {
 
     private val layers = TrieMap.empty[Int, Reader[SpatialKey, Tile]]
+    private val zoomLevels = 
+      for { LayerId(name, zoom) <- valueReader.attributeStore.layerIds if name == catalog } yield zoom
+    private val maxZoom = zoomLevels.max
 
     def retrieve(zoom: Int, x: Int, y: Int) = {
-      val key = SpatialKey(x, y)
       Future {
-        val reader = layers.getOrElseUpdate(zoom, valueReader.reader[SpatialKey, Tile](LayerId(catalog, zoom)))
-        Try(reader(key)).toOption
+        if (overzooming && zoom > maxZoom) {
+          val reader = layers.getOrElseUpdate(maxZoom, valueReader.reader[SpatialKey, Tile](LayerId(catalog, maxZoom)))
+          Try(rezoom(zoom, x, y, maxZoom, reader(_))).toOption
```

Comment:
Heads up, this `Try` is dangerous for debugging. You will want to `match` here for `Failure(e: ValueNotFoundError) => None`  and re-throw anything else.